internal/pkg/vcsutil/vcs_util.go

```diff
@@ -0,0 +1,276 @@
+package vcsutil
+
+import (
+	"fmt"
+	"os/exec"
+	"regexp"
+	"strings"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/hashicorp/go-hclog"
+	"github.com/pkg/errors"
+)
+
+var (
+	sshRemoteRegexp  *regexp.Regexp
+	httpRemoteRegexp *regexp.Regexp
+)
+
+func init() {
+	// Regex matching http/https remotes, tokenizing the unique components for replacement.
+	sshRemoteRegexp, _ = regexp.Compile(`git@(.*?):(.*\.git)`)                 // Example: git@git.test:testorg/testrepo.git
+	httpRemoteRegexp, _ = regexp.Compile(`http[s]?:\/\/(.*?\..*?)\/(.*\.git)`) // Example: https://git.test/testorg/testrepo.git
+}
+
+type VcsChecker struct {
+	log hclog.Logger
+
+	// path is the local path to the cloned git repo
+	path string
+
+	// remoteName is the name of the remote that corresponds to the url
+	remoteName string
+
+	// internal go-git repo
+	repo *git.Repository
+}
+
+// NewVcsChecker creates a new configured VcsChecker.
+// NOTE: VcsChecker subprocceses to the `git` command. Git must be installed.
+func NewVcsChecker(log hclog.Logger, path string, remoteUrl string) (*VcsChecker, error) {
```

Comment:
How do we plan on getting the "remote URL"? 

My recommendation would be to strike this, take an optional local branch, detect the [upstream tracking branch](https://git-scm.com/book/en/v2/Git-Branching-Remote-Branches) and then use that remote URL. That would be the way that git itself (if you run `git status`) detects if you have unpushed changes.

Reply:
I'm planning to use the remote url configured on the project settings. I'm reluctant to use tracking branches, because in many cases tracking isn't set at all, and if it is it's unlikely to directly track the remote/branch that we care about (i.e. the one the remote runner is going to use)